pkg/volume/cephfs/cephfs.go

```diff
@@ -0,0 +1,244 @@
+		mon:         ep,
+		secret:      spec.Cephfs.Secret,
+		readonly:    spec.Cephfs.ReadOnly,
+		id:          id,
+		secret_file: secrete_file,
+		mounter:     mounter,
+		exe:         exe,
+		podRef:      podRef,
+		plugin:      plugin,
+	}, nil
+}
+
+func (plugin *cephfsPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return plugin.newCleanerInternal(volName, podUID, mount.New())
+}
+
+func (plugin *cephfsPlugin) newCleanerInternal(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	return &cephfs{
+		volName: volName,
+		mounter: mounter,
+		podRef:  &api.ObjectReference{UID: podUID},
+		plugin:  plugin,
+	}, nil
+}
+
+// Cephfs volumes represent a bare host file or directory mount of an Cephfs export.
+type cephfs struct {
+	volName     string
+	podRef      *api.ObjectReference
+	mon         *api.Endpoints
+	id          string
+	secret      string
+	secret_file string
+	readonly    bool
+	mounter     mount.Interface
+	exe         exec.Interface
+	plugin      *cephfsPlugin
+}
+
+// SetUp attaches the disk and bind mounts to the volume path.
+func (cephfsVolume *cephfs) SetUp() error {
+	return cephfsVolume.SetUpAt(cephfsVolume.GetPath())
+}
+
+func (cephfsVolume *cephfs) SetUpAt(dir string) error {
+	mountpoint, err := cephfsVolume.mounter.IsMountPoint(dir)
+	glog.V(4).Infof("Cephfs: mount set up: %s %v %v", dir, mountpoint, err)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if mountpoint {
+		return nil
+	}
+	os.MkdirAll(dir, 0750)
+
+	err = cephfsVolume.execMount(dir)
+	if err == nil {
+		return nil
+	}
+
+	// cleanup upon failure
+	cephfsVolume.cleanup(dir)
+	// return error
+	return err
+}
+
+func (cephfsVolume *cephfs) GetPath() string {
+	name := cephfsPluginName
+	return cephfsVolume.plugin.host.GetPodVolumeDir(cephfsVolume.podRef.UID, util.EscapeQualifiedNameForDisk(name), cephfsVolume.volName)
+}
+
+func (cephfsVolume *cephfs) TearDown() error {
+	return cephfsVolume.TearDownAt(cephfsVolume.GetPath())
+}
+
+func (cephfsVolume *cephfs) TearDownAt(dir string) error {
+	return cephfsVolume.cleanup(dir)
+}
+
+func (cephfsVolume *cephfs) cleanup(dir string) error {
+	mountpoint, err := cephfsVolume.mounter.IsMountPoint(dir)
+	if err != nil {
+		glog.Errorf("Cephfs: Error checking IsMountPoint: %v", err)
+		return err
+	}
+	if !mountpoint {
+		return os.RemoveAll(dir)
+	}
+
+	if err := cephfsVolume.mounter.Unmount(dir, 0); err != nil {
+		glog.Errorf("Cephfs: Unmounting failed: %v", err)
+		return err
+	}
+	mountpoint, mntErr := cephfsVolume.mounter.IsMountPoint(dir)
+	if mntErr != nil {
+		glog.Errorf("Cephfs: IsMountpoint check failed: %v", mntErr)
+		return mntErr
+	}
+	if !mountpoint {
+		if err := os.RemoveAll(dir); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (cephfsVolume *cephfs) execMount(mountpoint string) error {
+	var errs error
+	var command exec.Cmd
+	var mountArgs []string
+	var opt []string
+
+	// cephfs mount option
```

Comment:
@vishh please use this container to test Ceph, https://github.com/ceph/ceph-docker/tree/master/demo